src/systems/audio-zones-system.js

```diff
@@ -0,0 +1,161 @@
+// We apply the most restrictive audio parameters
+function paramsReducer(acc, curr) {
+  if (acc === null) acc = curr;
+  acc.gain = Math.min(acc.gain, curr.gain);
+  acc.maxDistance = Math.min(acc.maxDistance, curr.maxDistance);
+  acc.refDistance = Math.min(acc.refDistance, curr.refDistance);
+  acc.rolloffFactor = Math.max(acc.rolloffFactor, curr.rolloffFactor);
+  acc.coneInnerAngle = Math.min(acc.coneInnerAngle, curr.coneInnerAngle);
+  acc.coneOuterAngle = Math.min(acc.coneOuterAngle, curr.coneOuterAngle);
+  acc.coneOuterGain = Math.min(acc.coneOuterGain, curr.coneOuterGain);
+  return acc;
+}
+
+/**
+ * This system updates the audio-zone-sources audio-params based on the audio-zone-listener position.
+ * On every tick it computes the audio-zone-source and audio-zone-listener positions to check
+ * if the listener is inside/outside an audio-zone and applies the zone's audio parameters based.
+ * It only updates in case there has been any changes in the sources or listener's positions.
+ * If several audio-zones are in between the audio-zone-listener and the audio-zone-source, the applied
+ * audio parameters is a reduction of the audio-zones most restrictive audio parameters.
+ * i.e. If there are two audio-zones in between the listener and the source and the first one has gain == 0.1
+ * and the other has gain == 1.0, gain == 0.1 is applied to the source.
+ */
+export class AudioZonesSystem {
+  constructor(scene) {
+    this.scene = scene;
+    this.listener = null;
+    this.sources = [];
+    this.zones = [];
+    this.initialized = false;
+    this.ray = new THREE.Ray();
+    this.rayDir = new THREE.Vector3();
+    this.normalizedRayDir = new THREE.Vector3();
+    this.intersectTarget = new THREE.Vector3();
+  }
+
+  remove() {
+    this.sources = [];
+    this.zones = [];
+  }
+
+  setListener(listener) {
+    this.listener = listener;
+  }
+
+  unsetListener() {
+    this.listener = null;
+  }
+
+  registerSource(source) {
+    this.sources.push(source);
+  }
+
+  unregisterSource(source) {
+    const index = this.sources.indexOf(source);
+
+    if (index !== -1) {
+      this.sources.splice(index, 1);
+    }
+  }
+
+  registerZone(zone) {
+    this.zones.push(zone);
+  }
+
+  unregisterZone(zone) {
+    const index = this.zones.indexOf(zone);
+
+    if (index !== -1) {
+      this.zones.splice(index, 1);
+    }
+  }
+
+  tick() {
+    if (!this.scene.is("entered") || this.zones.length === 0) return;
+
+    // Update zones
+    this._updateZones();
+
+    this.sources.forEach(source => {
```

Comment:
At some point we found that `forEach` created garbage so we stopped using it. No clue if this is still the case or just an urban legend at this point... but I tend to avoid it in hot code.

Reply:
I usually prefer the more functional style just for readability and code economy but for what I've read there is still some performance hit most noticeable with bigger arrays so I'm updating both this and the update zones one.